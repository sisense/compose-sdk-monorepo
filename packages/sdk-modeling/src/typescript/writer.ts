/* eslint-disable @typescript-eslint/no-throw-literal */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-argument */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/restrict-template-expressions */
import {
  BaseMeasure,
  DateDimension,
  Dimension,
  DimensionalDataModel,
  Element,
  MeasureTemplate,
  MetadataTypes,
} from '@sisense/sdk-data';
import { createWriteStream } from 'fs';
import path from 'path';

import { compileTsCode } from './utils/compile-ts-code.js';
import { createInMemoryDuplexStream } from './utils/create-in-memory-duplex-stream.js';
import { formatCode } from './utils/format-code.js';
import { ElementWriter, NEWLINE } from './writers/base.js';
import { DataSourceWriter } from './writers/datasource.js';
import { DateDimensionWriter, DimensionWriter } from './writers/dimensions.js';
import { ImportsWriter } from './writers/imports.js';
import { Writer } from './writers/interfaces.js';
import { BaseMeasureWriter, MeasureTemplateWriter } from './writers/measures.js';

function getMetadataWriter(item: Element): ElementWriter<any> {
  if (MetadataTypes.isDateDimension(item.type)) {
    return new DateDimensionWriter(<DateDimension>item, false);
  } else if (MetadataTypes.isDimension(item.type)) {
    return new DimensionWriter(<Dimension>item, false);
  } else if (MetadataTypes.isMeasureTemplate(item)) {
    return new MeasureTemplateWriter(<MeasureTemplate>item, false);
  } else if (MetadataTypes.isCalculatedMeasure(item)) {
    throw 'not implemented'; //return new SCalculatedMeasure(<CalculatedMeasure>item);
  } else if (MetadataTypes.isBaseMeasure(item)) {
    return new BaseMeasureWriter(<BaseMeasure>item, false);
  }

  throw 'unsupported metadata type';
}

type WriteConfig = {
  filename: string;
  dir?: string;
  datamodule?: string;
};

function fillStreamWithTsCode(stream: NodeJS.WritableStream, json: any, config: WriteConfig) {
  // generating a dimensional data model from the input json
  const dm = DimensionalDataModel.fromConfig(json);

  const writers: Writer[] = [];
  writers.push(new ImportsWriter(dm.dataSource, config.datamodule));
  writers.push(new DataSourceWriter(dm.dataSource));
  for (const metadataItem of dm.metadata) {
    writers.push(getMetadataWriter(metadataItem));
  }

  for (const writer of writers) {
    stream.write(NEWLINE);
    writer.write(stream, 0);
    stream.write(NEWLINE);
  }
  return stream;
}

export async function writeTypescript(json: any, config: WriteConfig) {
  if (!config.filename) {
    throw new Error('filename must be specified');
  }
  const filePath = path.join(config.dir || '', `${config.filename}.ts`);
  const fileStream = createWriteStream(filePath, { encoding: 'utf-8' });
  const tsCodeStream = createInMemoryDuplexStream();
  fillStreamWithTsCode(tsCodeStream, json, config);
  tsCodeStream.end();
  const unformattedTsCode = tsCodeStream.getStringData();
  const tsCode = await formatCode(unformattedTsCode, 'ts');
  fileStream.write(tsCode);
  fileStream.end();

  return new Promise<void>((resolve) => {
    fileStream.on('finish', () => {
      resolve();
    });
  });
}

export async function writeJavascript(json: any, config: WriteConfig) {
  if (!config.filename) {
    throw new Error('filename must be specified');
  }
  const tsCodeStream = createInMemoryDuplexStream();
  const jsFilePath = path.join(config.dir || '', `${config.filename}.js`);
  const dtsFilePath = path.join(config.dir || '', `${config.filename}.d.ts`);
  const jsFileStream = createWriteStream(jsFilePath, { encoding: 'utf-8' });
  const dtsFileStream = createWriteStream(dtsFilePath, { encoding: 'utf-8' });
  fillStreamWithTsCode(tsCodeStream, json, config);
  tsCodeStream.end();
  const tsCode = tsCodeStream.getStringData();
  const { jsCode: unformattedJsCode, typeDefs: unformattedTypeDefs } = compileTsCode(tsCode);
  const [jsCode, typeDefs] = await Promise.all([
    formatCode(unformattedJsCode, 'js'),
    formatCode(unformattedTypeDefs, 'ts'),
  ]);
  jsFileStream.write(jsCode);
  dtsFileStream.write(typeDefs);

  jsFileStream.end();
  dtsFileStream.end();

  return Promise.all([
    new Promise<void>((resolve) => {
      jsFileStream.on('finish', () => {
        resolve();
      });
    }),
    new Promise<void>((resolve) => {
      dtsFileStream.on('finish', () => {
        resolve();
      });
    }),
  ]);
}
