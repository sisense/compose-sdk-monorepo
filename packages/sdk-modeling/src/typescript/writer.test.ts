/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/* eslint-disable @typescript-eslint/no-unsafe-return */
import { writeJavascript, writeTypescript } from './writer.js';
import {
  dimensionalModelECommerce,
  generatedModelECommerce,
} from '../__mocks__/data-model-ecommerce.js';
import { Writable } from 'stream';

// A class for mocking fs.createWriteStream.
// It writes content to a buffer instead of to a file
// Adapted from https://bensmithgall.com/blog/jest-mock-trick
class WriteMemory extends Writable {
  buffer: string;

  constructor() {
    super();
    this.buffer = '';
  }

  _write(chunk: any, encoding: BufferEncoding, next: any) {
    this.buffer += chunk;
    next();
  }

  reset() {
    this.buffer = '';
  }
}

let mockWriteStream = new WriteMemory();

vi.mock('fs', async () => {
  const originalModule = await vi.importActual<typeof import('fs')>('fs');

  // Mock named export 'createWriteStream'
  return {
    __esModule: true,
    ...originalModule,
    createWriteStream: vi.fn().mockImplementation(() => mockWriteStream),
  };
});

function withoutSpaces(str: string) {
  return str.replace(/\s/g, '');
}

describe('writer', () => {
  beforeEach(() => {
    mockWriteStream = new WriteMemory();
  });
  describe('writeTypescript', () => {
    it('should generate data model file as TypeScript file', async () => {
      await writeTypescript(dimensionalModelECommerce, { filename: 'some-model' });
      const expectedCode = withoutSpaces(generatedModelECommerce.tsCode);
      const actualCode = withoutSpaces(mockWriteStream.buffer);
      expect(expectedCode).toEqual(actualCode);
    });
    it('should throw an error if filename is not specified', async () => {
      await expect(
        writeTypescript(dimensionalModelECommerce, {} as { filename: string }),
      ).rejects.toThrow('filename must be specified');
    });
  });

  describe('writeJavascript', () => {
    it('should generate JS and TS definition files', async () => {
      await writeJavascript(dimensionalModelECommerce, { filename: 'some-model' });

      const expectedJsCode = withoutSpaces(generatedModelECommerce.jsCode);
      const expectedDtsCode = withoutSpaces(generatedModelECommerce.dtsCode);
      const actualCode = withoutSpaces(mockWriteStream.buffer);
      // as we mocked the `fs` module with single stream, both .js and .d.ts files written to the same buffer
      expect(actualCode).toContain(expectedJsCode);
      expect(actualCode).toContain(expectedDtsCode);
    });

    it('should throw an error if filename is not specified', async () => {
      await expect(
        writeJavascript(dimensionalModelECommerce, {} as { filename: string }),
      ).rejects.toThrow('filename must be specified');
    });
  });
});
