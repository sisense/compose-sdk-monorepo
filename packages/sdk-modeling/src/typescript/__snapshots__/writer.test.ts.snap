// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`writer > writeJavascript > should generate JS and TS definition files 1`] = `
"import { createAttribute, createDateDimension, createDimension } from '@sisense/sdk-data';
export var DataSource = 'Sample ECommerce';
export var Brand = createDimension({
    name: 'Brand',
    Brand: createAttribute({
        name: 'Brand',
        type: 'text-attribute',
        expression: '[Brand.Brand]',
        description: 'Brand name',
    }),
    BrandID: createAttribute({
        name: 'Brand ID',
        type: 'numeric-attribute',
        expression: '[Brand.Brand ID]',
        description: 'Brand ID',
    }),
});
export var Category = createDimension({
    name: 'Category',
    Category: createAttribute({
        name: 'Category',
        type: 'text-attribute',
        expression: '[Category.Category]',
    }),
    CategoryID: createAttribute({
        name: 'Category ID',
        type: 'numeric-attribute',
        expression: "[Category.'Category ID']",
    }),
});
export var Commerce = createDimension({
    name: 'Commerce',
    AgeRange: createAttribute({
        name: 'Age Range',
        type: 'text-attribute',
        expression: '[Commerce.Age Range]',
    }),
    BrandID: createAttribute({
        name: 'Brand ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Brand ID]',
    }),
    CategoryID: createAttribute({
        name: 'Category ID',
        type: 'numeric-attribute',
        expression: "[Commerce.'Category ID']",
    }),
    Condition: createAttribute({
        name: 'Condition',
        type: 'text-attribute',
        expression: '[Commerce.Condition]',
    }),
    Cost: createAttribute({
        name: 'Cost',
        type: 'numeric-attribute',
        expression: '[Commerce.Cost]',
    }),
    CountryID: createAttribute({
        name: 'Country ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Country ID]',
    }),
    DateMonth: createAttribute({
        name: 'Date (Month)',
        type: 'numeric-attribute',
        expression: '[Commerce.Date (Month)]',
    }),
    Gender: createAttribute({
        name: 'Gender',
        type: 'text-attribute',
        expression: '[Commerce.Gender]',
    }),
    Quantity: createAttribute({
        name: 'Quantity',
        type: 'numeric-attribute',
        expression: '[Commerce.Quantity]',
    }),
    Revenue: createAttribute({
        name: 'Revenue',
        type: 'numeric-attribute',
        expression: '[Commerce.Revenue]',
    }),
    VisitID: createAttribute({
        name: 'Visit ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Visit ID]',
    }),
    Date: createDateDimension({
        name: 'Date',
        expression: '[Commerce.Date (Calendar)]',
    }),
});
export var Country = createDimension({
    name: 'Country',
    Country: createAttribute({
        name: 'Country',
        type: 'text-attribute',
        expression: '[Country.Country]',
    }),
    CountryID: createAttribute({
        name: 'Country ID',
        type: 'numeric-attribute',
        expression: '[Country.Country ID]',
    }),
});
import type { Dimension, DateDimension, Attribute } from '@sisense/sdk-data';
export declare const DataSource = 'Sample ECommerce';
interface BrandDimension extends Dimension {
    Brand: Attribute;
    BrandID: Attribute;
}
export declare const Brand: BrandDimension;
interface CategoryDimension extends Dimension {
    Category: Attribute;
    CategoryID: Attribute;
}
export declare const Category: CategoryDimension;
interface CommerceDimension extends Dimension {
    AgeRange: Attribute;
    BrandID: Attribute;
    CategoryID: Attribute;
    Condition: Attribute;
    Cost: Attribute;
    CountryID: Attribute;
    DateMonth: Attribute;
    Gender: Attribute;
    Quantity: Attribute;
    Revenue: Attribute;
    VisitID: Attribute;
    Date: DateDimension;
}
export declare const Commerce: CommerceDimension;
interface CountryDimension extends Dimension {
    Country: Attribute;
    CountryID: Attribute;
}
export declare const Country: CountryDimension;
export {};
"
`;

exports[`writer > writeTypescript > should generate data model file as TypeScript file (dataSource defined as DataSourceInfo) 1`] = `
"import type { Dimension, DateDimension, Attribute, DataSourceInfo } from '@sisense/sdk-data';

import { createAttribute, createDateDimension, createDimension } from '@sisense/sdk-data';

export const DataSource: DataSourceInfo = { title: 'Sample ECommerce', type: 'elasticube' };

interface BrandDimension extends Dimension {
    Brand: Attribute;
    BrandID: Attribute;
}
export const Brand = createDimension({
    name: 'Brand',
    Brand: createAttribute({
        name: 'Brand',
        type: 'text-attribute',
        expression: '[Brand.Brand]',
        description: \`Brand name\`,
    }),
    BrandID: createAttribute({
        name: 'Brand ID',
        type: 'numeric-attribute',
        expression: '[Brand.Brand ID]',
        description: \`Brand ID\`,
    }),
}) as BrandDimension;

interface CategoryDimension extends Dimension {
    Category: Attribute;
    CategoryID: Attribute;
}
export const Category = createDimension({
    name: 'Category',
    Category: createAttribute({
        name: 'Category',
        type: 'text-attribute',
        expression: '[Category.Category]',
    }),
    CategoryID: createAttribute({
        name: 'Category ID',
        type: 'numeric-attribute',
        expression: "[Category.'Category ID']",
    }),
}) as CategoryDimension;

interface CommerceDimension extends Dimension {
    AgeRange: Attribute;
    BrandID: Attribute;
    CategoryID: Attribute;
    Condition: Attribute;
    Cost: Attribute;
    CountryID: Attribute;
    DateMonth: Attribute;
    Gender: Attribute;
    Quantity: Attribute;
    Revenue: Attribute;
    VisitID: Attribute;
    Date: DateDimension;
}
export const Commerce = createDimension({
    name: 'Commerce',
    AgeRange: createAttribute({
        name: 'Age Range',
        type: 'text-attribute',
        expression: '[Commerce.Age Range]',
    }),
    BrandID: createAttribute({
        name: 'Brand ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Brand ID]',
    }),
    CategoryID: createAttribute({
        name: 'Category ID',
        type: 'numeric-attribute',
        expression: "[Commerce.'Category ID']",
    }),
    Condition: createAttribute({
        name: 'Condition',
        type: 'text-attribute',
        expression: '[Commerce.Condition]',
    }),
    Cost: createAttribute({
        name: 'Cost',
        type: 'numeric-attribute',
        expression: '[Commerce.Cost]',
    }),
    CountryID: createAttribute({
        name: 'Country ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Country ID]',
    }),
    DateMonth: createAttribute({
        name: 'Date (Month)',
        type: 'numeric-attribute',
        expression: '[Commerce.Date (Month)]',
    }),
    Gender: createAttribute({
        name: 'Gender',
        type: 'text-attribute',
        expression: '[Commerce.Gender]',
    }),
    Quantity: createAttribute({
        name: 'Quantity',
        type: 'numeric-attribute',
        expression: '[Commerce.Quantity]',
    }),
    Revenue: createAttribute({
        name: 'Revenue',
        type: 'numeric-attribute',
        expression: '[Commerce.Revenue]',
    }),
    VisitID: createAttribute({
        name: 'Visit ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Visit ID]',
    }),
    Date: createDateDimension({
        name: 'Date',
        expression: '[Commerce.Date (Calendar)]',
    }),
}) as CommerceDimension;

interface CountryDimension extends Dimension {
    Country: Attribute;
    CountryID: Attribute;
}
export const Country = createDimension({
    name: 'Country',
    Country: createAttribute({
        name: 'Country',
        type: 'text-attribute',
        expression: '[Country.Country]',
    }),
    CountryID: createAttribute({
        name: 'Country ID',
        type: 'numeric-attribute',
        expression: '[Country.Country ID]',
    }),
}) as CountryDimension;
"
`;

exports[`writer > writeTypescript > should generate data model file as TypeScript file 1`] = `
"import type { Dimension, DateDimension, Attribute } from '@sisense/sdk-data';

import { createAttribute, createDateDimension, createDimension } from '@sisense/sdk-data';

export const DataSource = 'Sample ECommerce';

interface BrandDimension extends Dimension {
    Brand: Attribute;
    BrandID: Attribute;
}
export const Brand = createDimension({
    name: 'Brand',
    Brand: createAttribute({
        name: 'Brand',
        type: 'text-attribute',
        expression: '[Brand.Brand]',
        description: \`Brand name\`,
    }),
    BrandID: createAttribute({
        name: 'Brand ID',
        type: 'numeric-attribute',
        expression: '[Brand.Brand ID]',
        description: \`Brand ID\`,
    }),
}) as BrandDimension;

interface CategoryDimension extends Dimension {
    Category: Attribute;
    CategoryID: Attribute;
}
export const Category = createDimension({
    name: 'Category',
    Category: createAttribute({
        name: 'Category',
        type: 'text-attribute',
        expression: '[Category.Category]',
    }),
    CategoryID: createAttribute({
        name: 'Category ID',
        type: 'numeric-attribute',
        expression: "[Category.'Category ID']",
    }),
}) as CategoryDimension;

interface CommerceDimension extends Dimension {
    AgeRange: Attribute;
    BrandID: Attribute;
    CategoryID: Attribute;
    Condition: Attribute;
    Cost: Attribute;
    CountryID: Attribute;
    DateMonth: Attribute;
    Gender: Attribute;
    Quantity: Attribute;
    Revenue: Attribute;
    VisitID: Attribute;
    Date: DateDimension;
}
export const Commerce = createDimension({
    name: 'Commerce',
    AgeRange: createAttribute({
        name: 'Age Range',
        type: 'text-attribute',
        expression: '[Commerce.Age Range]',
    }),
    BrandID: createAttribute({
        name: 'Brand ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Brand ID]',
    }),
    CategoryID: createAttribute({
        name: 'Category ID',
        type: 'numeric-attribute',
        expression: "[Commerce.'Category ID']",
    }),
    Condition: createAttribute({
        name: 'Condition',
        type: 'text-attribute',
        expression: '[Commerce.Condition]',
    }),
    Cost: createAttribute({
        name: 'Cost',
        type: 'numeric-attribute',
        expression: '[Commerce.Cost]',
    }),
    CountryID: createAttribute({
        name: 'Country ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Country ID]',
    }),
    DateMonth: createAttribute({
        name: 'Date (Month)',
        type: 'numeric-attribute',
        expression: '[Commerce.Date (Month)]',
    }),
    Gender: createAttribute({
        name: 'Gender',
        type: 'text-attribute',
        expression: '[Commerce.Gender]',
    }),
    Quantity: createAttribute({
        name: 'Quantity',
        type: 'numeric-attribute',
        expression: '[Commerce.Quantity]',
    }),
    Revenue: createAttribute({
        name: 'Revenue',
        type: 'numeric-attribute',
        expression: '[Commerce.Revenue]',
    }),
    VisitID: createAttribute({
        name: 'Visit ID',
        type: 'numeric-attribute',
        expression: '[Commerce.Visit ID]',
    }),
    Date: createDateDimension({
        name: 'Date',
        expression: '[Commerce.Date (Calendar)]',
    }),
}) as CommerceDimension;

interface CountryDimension extends Dimension {
    Country: Attribute;
    CountryID: Attribute;
}
export const Country = createDimension({
    name: 'Country',
    Country: createAttribute({
        name: 'Country',
        type: 'text-attribute',
        expression: '[Country.Country]',
    }),
    CountryID: createAttribute({
        name: 'Country ID',
        type: 'numeric-attribute',
        expression: '[Country.Country ID]',
    }),
}) as CountryDimension;
"
`;
